### PR TITLE
Support UNION_BY_NAME option in read_csv_auto 

### PR DESCRIPTION
**Pull request by douenergy**
https://github.com/duckdb/duckdb/issues/4720 (and https://github.com/duckdb/duckdb/issues/4699 )

Now, There is no need to write query like this  

```sql 
SELECT * FROM '[[]Avalon[]]_daily-avg.csv' UNION BY NAME
SELECT * FROM '[[]Liberty (SAHS)[]]_daily-avg.csv' UNION BY NAME
SELECT * FROM '[[]Parkway East (Near Road)[]]_daily-avg.csv' UNION BY NAME
SELECT * FROM '[[]Glassport High Street[]]_daily-avg.csv' UNION BY NAME
SELECT * FROM '[[]Manchester[]]_daily-avg.csv' UNION BY NAME
SELECT * FROM '[[]USA-Pennsylvania-Pittsburgh[]]_daily-avg.csv' UNION BY NAME
SELECT * FROM '[[]Lawrenceville[]]_daily-avg.csv' UNION BY NAME
SELECT * FROM '[[]North Braddock[]]_daily-avg.csv'
```
Just  😁

```sql 
SELECT * FROM  read_csv_auto('*.csv', UNION_BY_NAME=true)
```

All CSV will union by their header name.  If some CSV files without a specified header, read_csv_auto function will automatically add a Column filled with NULL. 

If different CSV files have some Column with the same name but different datatype, read_csv_auto  will choose the best datatype compatible with the original datatype

### CODE REVIEW

**Comment by douenergy** [comment]
I'll create another PR for UNION_BY_NAME option in read_parquet.

I think we need to check  what behavior we want for Parquet ?

Unlike CSV, Parquet format have exactly what datatype for every Columns.

So If we have p1.parquet 

```
col1
a
b
c
```
and p2.parquet 

```
col1
1
2
3
```

if we just cast col1 in  p2.parquet  to varchar 

 `INSERT INTO tbl *  SELECT * FROM read_parquet('*.parquet', FILENAME=TRUE);`

then 

`COPY (SELECT * FROM tbl WHERE filename='p2.parquet ') TO 'p2new.parquet';`

May surprised the user.  Because the col1 datatype change.

**Comment by djouallah** [comment]
if the same column has different type, I expect varchar, at least this is how Microsoft PowerQuery works

**Comment by douenergy** [comment]
cc @domoritz

**Comment by Mytherin** [comment]
Thanks for the PR! Very nice functionality.

Could you have a look at the failing CI runs? It is also possible to run and test the CI [locally in your fork](https://github.com/douenergy/duckdb/actions) if you enable Github actions runs there.

**Comment by douenergy** [comment]
Sure! I'll fix the failed test.  😊

**Comment by domoritz** [comment]
> if the same column has different type, I expect varchar, at least this is how Microsoft PowerQuery works

I feel that if the types differ in Parquet, there should be an error. Maaeeybe if the types are very similar (int32 and int64), the query could succeed with the more general type.

**Comment by djouallah** [comment]
> > if the same column has different type, I expect varchar, at least this is how Microsoft PowerQuery works
> 
> I feel that if the types differ in Parquet, there should be an error. Maaeeybe if the types are very similar (int32 and int64), the query could succeed with the more general type.

I feel strongly about this, it can be anything but definitely not errors, data quality is very bad out there, I would prefer to always load data then decide later what to do with it, not giving me errors :)

**Comment by domoritz** [comment]
I agree in general. My concern is that in parquet duckdb could read data efficiently without copies when the columns have the same type. Is the same true when the types differ?

**Comment by douenergy** [comment]
I think if two parquet columns have an incompatible type( read_parquet  currently cant cast ), we just cast them to varchar.  
Then read_parquet  warn the user, their parquet column type has changed.

What do you think? 

@Mytherin @djouallah @domoritz @lokax

**Comment by Mytherin** [comment]
I would use LogicalType::MaxLogicalType and the existing implicit cast rules to determine this, rather than always casting to varchar. If you have int32 and int64 you want to cast to int64, not to varchar. Perhaps having a flag to enable a more strict mode is also useful.

**Comment by douenergy** [comment]
Great 😄

 I think the read_parquet function  will have two options to control  union_by_name's behavior 
 `union_by_name`  
 `union_auto_cast` (default=false)
 
 `union_auto_cast` default is false, the default behavior is
   
 ** if implicit cast failed.  it will throw an error to the user** 

turn `union_auto_cast` to true , read_parquet  will use LogicalType::MaxLogicalType and the existing implicit cast rules to determine the proper type.

**Comment by domoritz** [comment]
I like the API for auto casting as described.

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good. Some comments:

File: test/sql/copy/csv/test_union_by_name.test [hunk 1]
```diff
@@ -0,0 +1,112 @@
+# name: test/sql/copy/csv/test_union_by_name.test
+# description: UNION_BY_NAME test
+# group: [csv]
+
+statement ok
```
Comment (Mytherin) [inline test/sql/copy/csv/test_union_by_name.test:5]: Could we also add a test that verifies that types are combined in a sane manner, e.g. `int32 + int64 = int64`, etc?

We could do this using the `typeof` function:

```sql
SELECT typeof(i::INT64) FROM range(10) tbl(i) LIMIT 1;
-- BIGINT
```

File: src/execution/operator/persistent/buffered_csv_reader.cpp [hunk 2]
```diff
@@ -1969,16 +1971,33 @@ bool BufferedCSVReader::AddRow(DataChunk &insert_chunk, idx_t &column) {
...
 
 	bool conversion_error_ignored = false;
+	const bool union_by_name = options.union_by_name;
 
 	// convert the columns in the parsed chunk to the types of the table
 	insert_chunk.SetCardinality(parse_chunk);
 	for (idx_t col_idx = 0; col_idx < sql_types.size(); col_idx++) {
+		idx_t insert_col_idx;
+		if (union_by_name) {
+			insert_col_idx = col_names_map[col_names[col_idx]];
```
[the remark this answers is not in the document]
Comment (douenergy) [reply src/execution/operator/persistent/buffered_csv_reader.cpp]: > Can't we create a `vector<idx_t>` that contains the remapping once per CSV reader in the bind phase and use that here?

I create `insert_cols_idx` for remapping

File: src/execution/operator/persistent/buffered_csv_reader.cpp [hunk 3]
```diff
@@ -1969,16 +1971,33 @@ bool BufferedCSVReader::AddRow(DataChunk &insert_chunk, idx_t &column) {
 	return false;
 }
 
+void BufferedCSVReader::SetNullUnionCols(DataChunk &insert_chunk, vector<string> &union_col_names) {
+	for (idx_t col = 0; col < union_col_names.size(); ++col) {
+		if (col_names_map.find(union_col_names[col]) == col_names_map.end()) {
```
[the remark this answers is not in the document]
Comment (douenergy) [reply src/execution/operator/persistent/buffered_csv_reader.cpp]: I create `insert_nulls_idx` for filling nulls

File: src/include/duckdb/execution/operator/persistent/buffered_csv_reader.hpp [hunk 4]
```diff
@@ -87,6 +87,10 @@ struct BufferedCSVReaderOptions {
 	FileCompressionType compression = FileCompressionType::AUTO_DETECT;
 	//! The column names of the columns to read/write
 	vector<string> names;
+	//! The header names of all csv files (option union_by_name)
+	vector<string> union_col_names;
```
[the remark this answers is not in the document]
Comment (douenergy) [reply src/include/duckdb/execution/operator/persistent/buffered_csv_reader.hpp]: I removed union_col_names now

File: test/sql/copy/csv/test_union_by_name.test [hunk 5]
```diff
@@ -0,0 +1,112 @@
+# name: test/sql/copy/csv/test_union_by_name.test
+# description: UNION_BY_NAME test
+# group: [csv]
```
[the remark this answers is not in the document]
Comment (douenergy) [reply test/sql/copy/csv/test_union_by_name.test:3]: I created an union_by_name_lineitem.test_slow  test

File: test/sql/copy/csv/test_union_by_name.test (hunk 1 above)
Comment (douenergy) [reply test/sql/copy/csv/test_union_by_name.test:5]: I add some new test for types check

**Comment by Mytherin** [comment]
Thanks for the updates! Looks great.

**Pull request merged**